File: struct-class/struct-class/stack.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// 구조체 정의
typedef struct Node{
    struct Node* NextNode;
    char Data[20];
} NODE;

typedef struct Stack{
    struct Node* Top;
    struct Node* Base;
} STACK;

// 선언
void InitStack(STACK** Stack);
NODE* CreateNode(char Data[]);
void Push(STACK* Stack, char Data[]);
NODE* Pop(STACK* Stack);
void printNode(NODE* Node);
int IsEmpty(STACK* Stack);

int main_stack(void)
{
    STACK* Stack;
    
    InitStack(&Stack);
    
    Push(Stack, "짜장면");
    Push(Stack, "짬뽕");
    Push(Stack, "군만두");
    printNode(Pop(Stack));
    printNode(Pop(Stack));
    Push(Stack, "탕수육");
    Push(Stack, "팔보채");
    
    printf("\n\n");
    while(1)
    {
        if(IsEmpty(Stack)) break; // 비어있으면 1, true출력, 안비었으면 0출력.
        printNode(Pop(Stack));
    }
    
    return 0;
}

// 정의
void InitStack(STACK** Stack)
{
    (*Stack) = (STACK*)malloc(sizeof(STACK));
    (*Stack)->Top = NULL;
    (*Stack)->Base = NULL;
}

NODE* CreateNode(char Data[])
{
    NODE* newNode = (NODE*)malloc(sizeof(NODE));
    strcpy(newNode->Data, Data);
    newNode->NextNode = NULL;
    return newNode;
}

void Push(STACK* Stack, char Data[])
{
    NODE* newNode = CreateNode(Data);
    if(Stack->Base == NULL){
        Stack->Base = newNode;
    } else{
        Stack->Top->NextNode = newNode;
    }
    Stack->Top = newNode;
}

NODE* Pop(STACK* Stack)
{
    NODE* deleteNode = Stack->Top;
    if(Stack->Base == NULL){
        return 1;
    } else if(Stack->Top == Stack->Base){
        Stack->Top = NULL;
        Stack->Base = NULL;
    } else {
        NODE* currNode = Stack->Base;
        while(currNode->NextNode != deleteNode && currNode != NULL){
            currNode = currNode->NextNode;
        }
        currNode->NextNode = NULL;
        Stack->Top = currNode;
    }
    return deleteNode;
}

void printNode(NODE* Node)
{
    printf("%s\n", Node->Data);
    free(Node);
}

int IsEmpty(STACK* Stack)
{
    return (Stack->Base == NULL);
}

File: js_bja/zerocho/stack/stack.js
class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class Stack {
  constructor() {
    this.node = null;
    this.length = 0;
  }

  push(data) {
    // 노드가 아예 없을 때
    if (!this.length) {
      this.length += 1;
      return (this.node = new Node(data));
    }
    // 하나라도 있으면 private메소드로 end point 탐색 후 삽입
    return this._setNode(this.node, data);
  }

  _setNode(node, data) {
    // 끝을 찾았을 때
    if (!node.next) {
      this.length += 1;
      return (node.next = new Node(data));
    }
    // 끝이 아닐 때는 재귀로 탐색
    return this._setNode(node.next, data);
  }

  pop() {
    return this._removeNode(this.node);
  }

  _removeNode(node) {
    if (!node.next.next) {
      const popValue = node.next.data;
      this.length -= 1;
      node.next = null;
      return popValue;
    }
    return this._removeNode(node.next);
  }
}

const stack = new Stack();
stack.push(2);
stack.push(4);
stack.push(7);
console.log(stack); // 2, 4, 7
console.log(stack.node.next);
console.log(stack.pop()); // 7
console.log(stack); // 2, 4
